syntax = "proto3";

import "google/protobuf/any.proto";
import "google/protobuf/struct.proto";

package example;

service Router {
    rpc Route1 (Request1) returns (Response) {}
    rpc Route2 (Request2) returns (Response) {}
}

// test field rule
message Request1 {
    // test repeated field rule
    repeated uint64 array = 1;
    // test oneof field rule
    oneof kind {
        double double_value = 2;
        string string_value = 3;
    }
}

enum SomeEnum  {
    UNIVERSAL = 0;
    WEB = 1;
    IMAGES = 2;
    LOCAL = 3;
    NEWS = 4;
    PRODUCTS = 5;
    VIDEO = 6;
}

message Bbox {
    uint64 left = 1;
    uint64 top = 2;
    uint64 right = 3;
    uint64 bottom = 4;
}

// test field type
message Request2 {
    // test bytes field type
    bytes b = 1;
    // test enum field type
    SomeEnum e = 2;
    // test map field type
    map<string, double> m = 3;
    // test Any field type
    google.protobuf.Any any = 4;
    // test NullValue field type
    google.protobuf.NullValue null = 5;
    // test ListValue field type
    google.protobuf.ListValue array = 6;
    // test nested message field type
    repeated Bbox bs = 7;
}

message Response {
    string text = 1;
}
